lib/upgrade/migrator/migrate.go

```diff
@@ -655,6 +657,62 @@ func (m *migrationHelper) queryAndConvertV1ToV3Nodes(data *MigrationData) error
 	return nil
 }
 
+// ShouldMigrate checks version information and reports if migration is needed
+// and is possible.
+func (m *migrationHelper) ShouldMigrate() (bool, error) {
+	_, err := m.clientv3.ClusterInformation().Get(context.Background(), "default", options.GetOptions{})
+	if err == nil {
+		// Nothing to do ClusterInformation exists so data has already been migrated
```

Comment:
Is that really a safe assertion?  The ClusterInformation is currently one of the first resources to be migrated across - it's possible that the migration could have failed part way through.

Reply:
IIUC I think we should be migrating if:
-  There is v3 cluster information and the cluster version is 2.x
-  There is no v3 cluster information

Checking the v1 cluster information doesn't tell us if we've already migrated - although it arguably tells us if we shouldn't migrate - e.g. because the cluster version is not set or is <2.6 (in which case they should be upgrading to the latest 2.6)

Reply:
Yes, I think you're right.